### PR TITLE
Monitoring JSON changes in DAQ (80X)

### PR DESCRIPTION
**Pull request by smorovic**
Changes:
- migrating to JSON based format for module legend used by f3mon
- adding HLTErrorEvents field which will used to distinguish between losses caused by crashes in CMSSW and other reasons, for example hardware failures.

port of #12136 submitted for 75X and #12137 submitted for 76X

### CODE REVIEW

**Comment by mommsen** [comment]
+1

**Comment by Martin-Grunewald** [comment]
please test

**Comment by fwyzard** [comment]
+1

**Comment by deguio** [comment]
+1

**Comment by Dr15Jones** [comment]
The failure is unrelated to this pull request

**Comment by slava77** [comment]
@cmsbuild please test
to get a cleaner test from jenkins (not including unrelated merges)

File: EventFilter/Utilities/src/FastMonitoringService.cc [hunk 1]
```diff
@@ -21,6 +21,10 @@ using namespace jsoncollector;
 
 constexpr double throughputFactor() {return (1000000)/double(1024*1024);}
 
+#define NRESERVEDMODULES 33
+#define NSPECIALMODULES 7
+#define NRESERVEDPATHS 1
```
Comment (cvuosalo) [inline EventFilter/Utilities/src/FastMonitoringService.cc]: Please don't use #define for constants. Please use "static const int" instead.

File: EventFilter/Utilities/src/FastMonitoringService.cc [hunk 2]
```diff
@@ -81,26 +85,34 @@ namespace evf{
...
-    std::ostringstream ost;
-    for(int i = 0;
-	i < encPath_[0].current_;
-	i++)
-      ost<<i<<"="<<*((std::string *)(encPath_[0].decode(i)))<<" ";
-    return ost.str();
+  std::string FastMonitoringService::makePathLegendaJson() {
+    Json::Value legendaVector(Json::arrayValue);
+    for(int i = 0; i < encPath_[0].current_; i++)
+      legendaVector.append(Json::Value(*((std::string *)(encPath_[0].decode(i)))));
```
Comment (cvuosalo) [inline EventFilter/Utilities/src/FastMonitoringService.cc]: The static analyzer has flagged this cast. I think the cast should be (const std::string *).

**Comment by smorovic** [comment]
@cvuosalo, thanks for comments. It has been addressed in latest commit.

**Comment by mommsen** [comment]
+1

**Comment by slava77** [comment]
there are merge conflicts here now.
this is most likely after the auto-forward-port of #12137

File: EventFilter/Utilities/src/FastMonitoringService.cc [hunk 3]
```diff
@@ -81,26 +85,34 @@ namespace evf{
...
-
-  std::string FastMonitoringService::makeModuleLegenda()
-  {  
-    std::ostringstream ost;
+  std::string FastMonitoringService::makeModuleLegendaJson(){
+    Json::Value legendaVector(Json::arrayValue);
     for(int i = 0; i < encModule_.current_; i++)
-      ost<<i<<"="<<((const edm::ModuleDescription *)(encModule_.decode(i)))->moduleLabel()<<" ";
-    return ost.str();
+       legendaVector.append(Json::Value(((const edm::ModuleDescription *)(encModule_.decode(i)))->moduleLabel()));
```
Comment (Dr15Jones) [inline EventFilter/Utilities/src/FastMonitoringService.cc]: Please avoid use of C style casts. The C++ casts are much easier to notIce and avoid accidental casting away can const.

File: EventFilter/Utilities/src/FastMonitoringService.cc [hunk 4]
```diff
@@ -81,26 +85,34 @@ namespace evf{
...
-    std::ostringstream ost;
-    for(int i = 0;
-	i < encPath_[0].current_;
-	i++)
-      ost<<i<<"="<<*((std::string *)(encPath_[0].decode(i)))<<" ";
-    return ost.str();
+  std::string FastMonitoringService::makePathLegendaJson() {
+    Json::Value legendaVector(Json::arrayValue);
+    for(int i = 0; i < encPath_[0].current_; i++)
+      legendaVector.append(Json::Value(*((const std::string *)(encPath_[0].decode(i)))));
```
Comment (Dr15Jones) [inline EventFilter/Utilities/src/FastMonitoringService.cc]: Please use C++ style cast.

**Comment by smorovic** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
@smorovic 
there are merge conflicts in this PR. So, it can not proceed without the merge issues resolved.
Please rebase this to a more recent 80X IB (anything starting from CMSSW_8_0_X_2015-10-30-2300) and re-test there.

**Comment by smorovic** [comment]
rebase got messed up. Will open another PR and branch for this.

**Pull request closed without merging**